Enable Block Email Editor Support for Automation Newsletters - STOMAIL-7541

triple0t opened the pull request:
## Description

This PR introduces comprehensive support for using the modern block email editor (WooCommerce Email Editor) when creating and editing automation newsletters in MailPoet. This enhancement provides users with a more modern, flexible, and powerful email editing experience for automation workflows.

## Why

Previously, automation newsletters were limited to the legacy MailPoet newsletter editor, which lacked the modern block-based editing capabilities available to other email types. This created an inconsistent user experience and prevented automation emails from leveraging the enhanced design capabilities of the block editor.

The block email editor offers:
- Modern block-based interface with better UX
- Enhanced design flexibility and customization options
- Consistent editing experience across different email types
- Future-proof foundation for email editing features

### Key Changes

1. **Settings Integration**
   - Added new setting "Use block email editor for automation newsletter" in Advanced settings
   <img width="1052" height="134" alt="Screenshot 2025-09-01 at 12 56 37 PM" src="https://github.com/user-attachments/assets/187972c8-fc87-47ef-a994-f01dce58fc65" />

2. **Editor Integration**
   - Enhanced `EditNewsletter` component to detect and redirect to the appropriate editor
   - Added support for `wp_post_id` tracking for block editor emails
   - Implemented proper URL generation for both legacy and block editors

3. **New filter**
- Add filter `mailpoet_use_block_email_editor_for_automation_emails` to control the feature.
- Returning `false` will disable the feature.

4. **UI/UX Improvements**
   - Context-aware editor labels ("Save and continue" for automation emails)
   - Hidden unnecessary fields (subject/preheader) in block editor for automation emails
   - Proper navigation and URL handling between automation editor and email editor

5. **Infrastructure**
   - Centralized URL generation utilities (`getNewsletterEditorUrl`, `getTheEmailEditorUrl`)
   - Enhanced cleanup logic for unused automation steps
   - Comprehensive acceptance test coverage

## Code review notes

_N/A_

## QA notes

**UI Mode**
1. Enable the feature in `/wp-admin/admin.php?page=mailpoet-settings#/advanced`, 
2. Scroll down to `Use block email editor for automation newsletter`, select `Yes`, and click on "Save settings"
3. Create a new automation in `/wp-admin/admin.php?page=mailpoet-automation`
4. Ensure you have the "Send email" step, click on design, notice you are redirected to the block email page
5. Design an email, save, and activate the automation.

**Filter Mode**
1. Add filter for `mailpoet_use_block_email_editor_for_automation_newsletter` (code below)
2. Create a new automation in `/wp-admin/admin.php?page=mailpoet-automation`
3. Ensure you have the "Send email" step, click on design, notice you are redirected to the block email page
4. Design an email, save, and activate the automation.

```php
add_filter('mailpoet_use_block_email_editor_for_automation_emails', '__return_true', 10, 1);
```

## Linked PRs

_N/A_

## Linked tickets

[STOMAIL-7541](https://linear.app/a8c/issue/STOMAIL-7541/spike-on-using-the-block-email-editor-for-creating-automations-emails)

## After-merge notes

_N/A_

## Tasks

- [ ] I have added a changelog entry (`./do changelog:add --type=<type> --description=<description>`)
- [ ] I followed [best practices](https://codex.wordpress.org/I18n_for_WordPress_Developers) for translations
- [ ] I added sufficient test coverage
- [ ] I embraced TypeScript by either creating new files in TypeScript or converting existing JavaScript files when making changes

## Preview

[Preview in WordPress Playground](https://account.mailpoet.com/playground/new/branch:stomail-7541-spike-on-using-the-block-email-editor-for-automation-emails)

_The latest successful build from `stomail-7541-spike-on-using-the-block-email-editor-for-automation-emails` will be used. If none is available, the link won't work._

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- New Features
  - Enable creating and editing automation emails with the block email editor; edit/preview links now open the appropriate editor and block-editor UI hides subject/preview fields.
- Settings
  - Advanced settings toggle to choose block editor for automation newsletters; translations and a global flag added.
- Refactor
  - Centralized editor URL generation and exposed feature flag to drive editor selection.
- Bug Fixes
  - Prevent removal of emails tied to WP posts with block-editor content.
- Tests
  - Add end-to-end tests covering block-editor enabled/disabled automation flows.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

triple0t committed:
Refactor new editor check in NewsletterSaveController

Ensure we check the value of the `new_editor` variable ('true' or 'false') to confirm creating a new WP post for the newsletter

Previously, passing 'false' to the new_editor property would still create the newsletter WP_Post instead of skipping it.

STOMAIL-7541

triple0t committed:
Enhance newsletter deletion logic to skip posts with content from the new block editor. Added checks for existing send email steps to remove unused `email_wp_post_id` value

STOMAIL-7541

triple0t committed:
Add newsletter editor URL functions to MailPoet

Introduced `getNewsletterEditorUrl` and `getTheEmailEditorUrl` functions to generate URLs for the newsletter editor based on the newsletter ID and context. Updated the MailPoet object to include these new functions.

STOMAIL-7541

triple0t committed:
Update Send email step (EditNewsletter) - Allow using block editor for Automation emails

This will create newsletters using the block email editor and allow redirecting to the email editor page.

STOMAIL-7541

triple0t committed:
Add integration for Automation newsletters in email editor

- Introduced TypeScript definitions for WooCommerceEmailEditor in global.d.ts.
- Updated index.ts to conditionally display custom labels and hide unnecessary fields for Automation newsletters.
- Modified EditorPageRenderer.php to handle URLs based on whether the newsletter is an Automation type.
- Added methods in NewsletterEntity.php to check if a newsletter is Automation or Automation Transactional.

STOMAIL-7541

triple0t committed:
Enhance email actions to support editing Automation newsletters created with the block email editor.

- Added `wpPostId` to the `Actions` component to better determine which editor to open and to allow using the block email editor.
- Updated `transformEmailsToRows` to pass `wpPostId` to the `Actions` component.
- Modified `EmailStats` type to include optional `wpPostId`.
- Updated `OverviewStatisticsController` to include `wpPostId` in the email data.

STOMAIL-7541

triple0t committed:
Refactor newsletter editor URL generation to use unified method (sse Global helper methods instead of hardcoded strings)

- Updated `getEditorLink` function to utilize `MailPoet.getTheEmailEditorUrl` for consistent URL generation across various components.
- Modified links in `notification.jsx`, `re-engagement.jsx`, `standard.jsx`, and `utils.jsx` to replace direct URL construction with the new method.

STOMAIL-7541

triple0t committed:
Switch to dynamic data instead of the hardcoded boolean value for using the block email editor for automation newsletter

STOMAIL-7541

triple0t committed:
Add dynamic support for block email editor in automation newsletters

- Introduced a new method `useBlockEmailEditorForAutomationNewsletter` in `PageRenderer` to retrieve the setting for using the block email editor.
- Updated `layout.html` to include the new dynamic variable for the block email editor setting.

STOMAIL-7541

triple0t committed:
Add Settings UI for using the block editor for automation newsletters

STOMAIL-7541

triple0t committed:
Add changelog file

triple0t committed:
Add acceptance tests for using the block editor for automation newsletter

- Added automation IDs to the block editor toggle options in the settings page for better accessibility.
- Created a new acceptance test class to verify the functionality of creating automation emails using the block editor, including enabling/disabling the block editor setting and ensuring the correct editor is used based on the setting.

STOMAIL-7541

triple0t committed:
TIny fixes: review feedback

STOMAIL-7541

lysyjan commented on mailpoet/assets/js/src/mailpoet.ts:
```diff
@@ -93,6 +117,8 @@ export const MailPoet = {
   cronTriggerMethod: window.mailpoet_cron_trigger_method,
   getBlockEmailEditorUrl,
+  getNewsletterEditorUrl,
+  getTheEmailEditorUrl,
```
Would it be better to use the name `getEmailEditorUrl` only? I think it's not common to use an article in names.

triple0t replied:
Because `getEmailEditorUrl` can easily be confused for the block email editor URL. 
I would rather not even use it for this use case. 
Do you have another idea of a good name for this method?

lysyjan replied:
What about `getActiveEmailEditorUrl()`? Or something like provider?

triple0t replied:
Let's go with `getActiveEmailEditorUrl`

Thank you.

triple0t replied:
Implemented in https://github.com/mailpoet/mailpoet/pull/6314/commits/79f60dc077209fa8dbb3134960ac2446a368bd52

lysyjan commented on mailpoet/assets/js/src/automation/integrations/mailpoet/analytics/components/tabs/emails/rows.tsx:
```diff
@@ -131,7 +131,13 @@ export function transformEmailsToRows(emails: EmailStats[]) {
+          <Actions
+            id={email.id}
+            previewUrl={email.previewUrl}
+            wpPostId={email?.wpPostId ?? 0}
```
When the post ID is not set, it is more common to use a value of type `null`, `undefined`, or `false` than zero. Also, zero can be a valid ID. I know it is not typical, but it's possible. Or is there something I am missing?

triple0t replied:
Zero is considered as falsey in both JavaScript and [PHP](https://www.php.net/manual/en/language.types.boolean.php).
<img width="633" height="71" alt="Screenshot 2025-09-09 at 10 33 11 AM" src="https://github.com/user-attachments/assets/ba4ad9e9-0a32-4ce9-9950-64b05ac2b741" />
<img width="877" height="211" alt="Screenshot 2025-09-09 at 10 35 46 AM" src="https://github.com/user-attachments/assets/4383b0b6-72e2-42d7-b4c0-f3a62d43b309" />

WordPress also [does not consider](https://developer.wordpress.org/reference/functions/get_post/) zero as a valid post ID.

triple0t replied:
Updated in https://github.com/mailpoet/mailpoet/pull/6314/commits/4ac034298a44dfb5922dbf6a864796fbcc1c02ab

lysyjan commented on mailpoet/assets/js/src/automation/integrations/mailpoet/analytics/store/types.ts:
```diff
@@ -21,6 +21,7 @@ export type EmailStats = {
   orders: number;
   revenue: number;
   unsubscribed: number;
+  wpPostId?: number;
```
This relates to the previous comment and coderabbitai's suggestion. I would say we do not have to cast into `int`, but we could use `number | null`. What do you think?

triple0t replied:
Updated in https://github.com/mailpoet/mailpoet/pull/6314/commits/4ac034298a44dfb5922dbf6a864796fbcc1c02ab

lysyjan commented on mailpoet/assets/js/src/global.d.ts:
```diff
@@ -207,6 +207,7 @@ interface Window {
   mailpoet_installed_days_ago: number;
   mailpoet_send_transactional_emails: boolean;
   mailpoet_transactional_emails_opt_in_notice_dismissed: boolean;
+  mailpoet_use_block_email_editor_for_automation_newsletter: boolean;
```
The suffix `newsletter` does not make sense here. If I am not wrong, `newsletter` is a regular email. But automation can also send transactional emails. I think ` mailpoet_use_block_email_editor_for_automation` would be enough. What do you think?

triple0t replied:
Will switch to `mailpoet_use_block_email_editor_for_automation_emails`

triple0t replied:
Updated in https://github.com/mailpoet/mailpoet/pull/6314/commits/4ac034298a44dfb5922dbf6a864796fbcc1c02ab

lysyjan commented on mailpoet/views/settings_translations.html:
```diff
@@ -107,6 +107,8 @@
   'after18months': __('After 18 months'),
   'libs3rdPartyTitle': __('Load 3rd-party libraries'),
   'libs3rdPartyDescription': __('E.g. Google Fonts in the Form and Email editor and WordPress.com to get help. When disabled, you can still reach support at [link]www.mailpoet.com/support/[/link].'),
+  'useBlockEditorForAutomationTitle': __('Use block email editor for automation newsletter'),
```
I am not sure about the title. Maybe we can ask Shiva for help here.
My suggestion would be:
`Use block email editor in Automations feature`.

triple0t replied:
Updated in https://github.com/mailpoet/mailpoet/pull/6314/commits/4ac034298a44dfb5922dbf6a864796fbcc1c02ab

lysyjan reviewed (commented):
@triple0t Good job. Overall, it looks good to me. I am just not sure about some terminology and new texts in the settings.

triple0t committed:
PR feedback
Replace `use_block_email_editor_for_automation_newsletter` with `use_block_email_editor_for_automation_emails`

STOMAIL-7541

triple0t committed:
PR feedback - replace `getTheEmailEditorUrl` with `getActiveEmailEditorUrl`

and Fix some typo

lysyjan commented on mailpoet/views/settings_translations.html:
```diff
@@ -107,7 +107,7 @@
   'libs3rdPartyTitle': __('Load 3rd-party libraries'),
   'libs3rdPartyDescription': __('E.g. Google Fonts in the Form and Email editor and WordPress.com to get help. When disabled, you can still reach support at [link]www.mailpoet.com/support/[/link].'),
-  'useBlockEditorForAutomationTitle': __('Use block email editor for automation newsletter'),
+  'useBlockEditorForAutomationTitle': __('Use block email editor for automation eamils'),
```
@triple0t Thanks for the update. I noticed this typo.

triple0t replied:
Yeah, the typo was already fixed in https://github.com/mailpoet/mailpoet/pull/6314/commits/79f60dc077209fa8dbb3134960ac2446a368bd52

lysyjan reviewed (approved):
Thanks. It looks good 👍

Aschepikov reviewed (approved):
GJ!
